scanpy/tools/_score_genes.py

```diff
@@ -155,12 +155,24 @@ def score_genes(
     control_genes = pd.Index([], dtype="string")
 
     # now pick `ctrl_size` genes from every cut
-    for cut in np.unique(obs_cut.loc[gene_list]):
-        r_genes: pd.Index[str] = obs_cut[obs_cut == cut].index
+    for cut in obs_cut.loc[gene_list].unique():
+        r_genes: pd.Index[str] = obs_cut[
+            (obs_cut == cut) & ~obs_cut.index.isin(gene_list)
+        ].index
+        if len(r_genes) == 0:
+            logg.warning(
+                f"No control genes for this cut. You might want to increase "
+                f"ctrl_size={ctrl_size} to sample from more control genes."
+            )
         if ctrl_size < len(r_genes):
             r_genes = r_genes.to_series().sample(ctrl_size).index
         control_genes = control_genes.union(r_genes.difference(gene_list))
```

Comment:
> So if the original implementation does not remove score genes from the control gene set, we would simply need to remove the following line:
> https://github.com/scverse/scanpy/blob/ec4457470618efd85da3c7b29f951cab01a49e3a/scanpy/tools/_score_genes.py#L169
> (Note: if we want to keep the current behaviour, we should still remove the line above, since it would be redundant)

Removing that line completely wouldn’t make sense, then `control_genes` would be empty. But with your `& ~obs_cut.index.isin(gene_list)` change, a part of the line currently redundant, so we could do this without changing behavior:

```suggestion
        control_genes = control_genes.union(r_genes)
```

Reply:
ah yes, you're right, it's just the difference that should be redundant